lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp

```diff
@@ -1343,14 +1372,56 @@ struct LowerTypesPass : public LowerFIRRTLTypesBase<LowerTypesPass> {
 // This is the main entrypoint for the lowering pass.
 void LowerTypesPass::runOnOperation() {
   std::vector<Operation *> ops;
-  llvm::for_each(getOperation().getBody()->getOperations(),
-                 [&](Operation &op) { ops.push_back(&op); });
+  // Record of each module, to the map of NonLocalAnchors which need to be
+  // updated with the lowered inner_sym name.
+  DenseMap<Operation *, DenseMap<StringAttr, StringAttr>>
+      modToNlaNameToNewSymMap;
+  // Map of name of the NonLocalAnchor to the operation.
+  DenseMap<StringAttr, Operation *> nlaMap;
+  // Record all operations in the circuit.
+  llvm::for_each(getOperation().getBody()->getOperations(), [&](Operation &op) {
+    ops.push_back(&op);
+    // Creating a map of all ops in the circt, but only modules are relevant.
+    modToNlaNameToNewSymMap.insert(
+        std::make_pair<Operation *, DenseMap<StringAttr, StringAttr>>(
+            &op, /*An empty map*/ DenseMap<StringAttr, StringAttr>()));
+    // Record the NonLocalAnchor and its name.
+    if (auto nla = dyn_cast<NonLocalAnchor>(op))
+      nlaMap[nla.sym_nameAttr()] = nla;
+  });
```

Comment:
This pattern may not be thread safe as each thread may cause the whole `DenseMap` to get rebuilt.  I think you're looking for a `parallelTransformReduce` pattern here where the maps get merged at the end. That said, tracking the outer `Operation *` (module) actually necessary here? This is really trying to track updates the the leaf of an NLA and that can only exist for each NLA.

I think you can get away with passing one `DenseMap<StringAttr, StringAttr>` to each thread and then merging them at the end.

Reply:
Used the reduce pattern, and replaced the maps with a single `SmallVector`, because the NLA list is anyway unique.
There is a one to one correspondence between the leaf element symbol and the corresponding NLA.